### PR TITLE
Port the auto tag feature from docker plugin

### PR DESCRIPTION
**Pull request by 15cm**
Support the feature in the ["Autotag" section](https://plugins.drone.io/drone-plugins/drone-docker/) of docker plugin.

The logic is forked from https://github.com/drone-plugins/drone-docker code base with necessary modification. I've tested it e2e for DockerHub on my Drone server via this plugin image https://hub.docker.com/repository/docker/15cm/drone-kaniko, for both tag pushes and commit pushes.

With this change the .drone.yml in this repo should work as intended.

Other changes:
- Rename the existing "auto tag" flags/code to "expand tag" for a less misleading naming.
- ATTENTION: make a breaking change to set default value of "--tags" to empty.
  Rationale is to expect most users to use the auto tagging feature. When power users want to specify tags, they should always explicitly set tags instead of being surprised by the default "latest" tag.

### CODE REVIEW

**Comment by shubham149** [comment]
We cannot make any breaking changes until and unless they are security vulnerabilities. Auto tag was already supported in this plugin. Can you update what functionality was earlier missing for auto-tags?

**Comment by 15cm** [comment]
The existing auto tag feature behaves differently from the auto tag feature of the [drone docker plugin](https://plugins.drone.io/drone-plugins/drone-docker/). Comparison:
- The current auto tag requires users to specify the tag in the CLI argument. It expands the tag into several tags following semantic versioning. 
- The new auto tag in this PR automatically detect the tag without a user provided tag. Then expand it following semantic versioning.

The reason I make breaking change on the AUTO_TAG flag instead of creating a new flag is that, the current auto tag is not "auto". I have to specify the tag, which means an update of .drome.yml per tag push. Moreover, as a user who comes from the drone docker plugin (I believe I'm not the only one) that supports the auto tag feature that behaves the same as this PR, I was surprised by the behavior of the current auto tag feature and struggled to see if it could auto detect the tag of git commits.

In summary, I think the current auto tag feature:
- diverges from the drone docker plugin which will surprise many users.
- doesn't suit the AUTO_TAG name well given that we have a way to auto detect the tag from git commit (what I do in this PR).

File: cmd/kaniko-gcr/main.go [hunk 1]
```diff
@@ -155,12 +176,19 @@ func run(c *cli.Context) error {
 		}
 	}
 
+	enableAutoTag := c.Bool("tags.auto")
+	enableExpandTag := c.Bool("tags.expand")
+	tags, shouldSkipBuild, err := tagger.MaybeAutoTag(c.StringSlice("tags"), c.String("commit.ref"), c.String("tags.auto_suffix"), c.String("repo.branch"), enableAutoTag, enableExpandTag)
```
Comment (shubham149) [inline cmd/kaniko-gcr/main.go:181]: If auto-tag is not enabled, then tags will be empty list which is not correct. Also, L179-L181 should be moved to kaniko.go file since it is same for all of dockerhub, gcr & ecr.

**Review by shubham149 — COMMENTED** [review]
@15cm Sounds good. I have added a comment to the PR.

File: cmd/kaniko-gcr/main.go (hunk 1 above)
Comment (15cm) [reply cmd/kaniko-gcr/main.go]: Limited the breaking changes to users that enable the current auto-tag flag. Behaviors comparison before/after the PR:
* --auto-tag=false: No changes.
* --auto-tag=false,--expand-tag=true,tags=1.0.0:
  * Old behavior: Should not happen. --expand-tag didn't exist.
  * New Behavior: Build with [1,1.0,1.0.0] tags.
* --auto-tag=true
  * Old behavior: Build with the "latest" tag.
  * New behavior: Build with auto detected tags. Abort if auto detection failed.
* --auto-tag=true,tags=latest: same as "--auto-tag=true".
* --auto-tag=true,tags=1.0.0:
  * Old behavior: Build with [1,1.0,1.0.0] tags.
  * New behavior: Abort the build with an error message.
* --auto-tag=true,--expand-tag=true,tags=1.0.0: Abort the build with an error message.

Moved the logic of L179-L181 to Plugin::Exec() in "kaniko.go".

**Review by shubham149 — APPROVED** [review]

**Comment by shubham149** [comment]
Thanks a lot @15cm for your contribution.

**Comment by 15cm** [comment]
Thanks for taking care of this PR @shubham149. Sorry a link in the README added by this PR is malformatted. I created another PR as a minor fix https://github.com/drone/drone-kaniko/pull/38.
